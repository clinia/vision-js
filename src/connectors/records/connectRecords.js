import {
  checkRendering,
  createDocumentationMessageGenerator,
  addAbsolutePosition,
  addQueryID,
  noop,
} from '../../lib/utils';

const withUsage = createDocumentationMessageGenerator({
  name: 'records',
  connector: true,
});

/**
 * @typedef {Object} RecordsRenderingOptions
 * @property {Object[]} records The matched records from Clinia API.
 * @property {Object} results The complete results response from Clinia API.
 * @property {Object} widgetParams All original widget options forwarded to the `renderFn`.
 */

/**
 * @typedef {Object} CustomRecordsWidgetOptions
 * @property {function(Object[]):Object[]} [transformItems] Function to transform the items passed to the templates.
 */

/**
 * **Records** connector provides the logic to create custom widgets that will render the results retrieved from Clinia.
 * @type {Connector}
 * @param {function(RecordsRenderingOptions, boolean)} renderFn Rendering function for the custom **Records** widget.
 * @param {function} unmountFn Unmount function called when the widget is disposed.
 * @return {function(CustomRecordsWidgetOptions)} Re-usable widget factory for a custom **Records** widget.
 * @example
 * // custom `renderFn` to render the custom Records widget
 * function renderFn(RecordsRenderingOptions) {
 *   RecordsRenderingOptions.widgetParams.containerNode.html(
 *     RecordsRenderingOptions.records.map(function(record) {
 *       return '<div>' + records._highlightResult.name.value + '</div>';
 *     })
 *   );
 * }
 *
 * // connect `renderFn` to Records logic
 * var customRecords = vision.connectors.connectRecords(renderFn);
 *
 * // mount widget on the page
 * search.addWidgets([
 *   customRecords({
 *     containerNode: $('#custom-records-container'),
 *   })
 * ]);
 */
export default function connectRecords(renderFn, unmountFn = noop) {
  checkRendering(renderFn, withUsage());

  return (widgetParams = {}) => {
    const { transformItems = items => items } = widgetParams;

    return {
      $$type: 'cvi.records',

      init({ visionInstance }) {
        renderFn(
          {
            records: [],
            results: undefined,
            visionInstance,
            widgetParams,
          },
          true
        );
      },

      render({ results, visionInstance }) {
        const initialEscaped = results.records.__escaped;

        results.records = addAbsolutePosition(
          results.records,
          results.page,
          results.perPage
        );

        results.records = addQueryID(results.records, results.queryId);

        results.records = transformItems(results.records);

        // Make sure the escaped tag stays, even after mapping over the records.
        // This prevents the records from being double-escaped if there are multiple
        // records widgets mounted on the page.
        results.records.__escaped = initialEscaped;

        renderFn(
          {
            records: results.records,
            results,
            visionInstance,
            widgetParams,
          },
          false
        );
      },

      dispose({ state }) {
        unmountFn();

        return state;
      },

      getWidgetSearchParameters(state) {
        return state;
      },
    };
  };
}
