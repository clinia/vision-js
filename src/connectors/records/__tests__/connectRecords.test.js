import searchHelper, {
  SearchParameters,
  SearchResults,
} from '@clinia/search-helper';
import connectRecords from '../connectRecords';

jest.mock('../../../lib/utils/records-absolute-position', () => ({
  // The real implementation creates a new array instance, which can cause bugs,
  // especially with the __escaped mark, we thus make sure the mock also has the
  // same behavior regarding the array.
  addAbsolutePosition: records => records.map(x => x),
}));

describe('connectRecords', () => {
  it('throws without render function', () => {
    expect(() => {
      connectRecords()({});
    }).toThrowErrorMatchingSnapshot();
  });

  it('is a widget', () => {
    const render = jest.fn();
    const unmount = jest.fn();

    const customRecords = connectRecords(render, unmount);
    const widget = customRecords({});

    expect(widget).toEqual(
      expect.objectContaining({
        $$type: 'cvi.records',
        init: expect.any(Function),
        render: expect.any(Function),
        dispose: expect.any(Function),
      })
    );
  });

  it('Renders during init and render', () => {
    const renderFn = jest.fn();
    const makeWidget = connectRecords(renderFn);
    const widget = makeWidget({ escapeHTML: true });

    // test if widget is not rendered yet at this point
    expect(renderFn).toHaveBeenCalledTimes(0);

    const helper = searchHelper({}, '', {});
    helper.search = jest.fn();

    widget.init({
      helper,
      state: helper.state,
      createURL: () => '#',
    });

    expect(renderFn).toHaveBeenCalledTimes(1);
    expect(renderFn).toHaveBeenLastCalledWith(
      expect.objectContaining({ widgetParams: { escapeHTML: true } }),
      true
    );

    widget.render({
      results: new SearchResults(helper.state, [{ records: [], meta: {} }]),
      state: helper.state,
      helper,
      createURL: () => '#',
    });

    expect(renderFn).toHaveBeenCalledTimes(2);
    expect(renderFn).toHaveBeenLastCalledWith(
      expect.objectContaining({ widgetParams: { escapeHTML: true } }),
      false
    );
  });

  it('Provides the records and the whole results', () => {
    const renderFn = jest.fn();
    const makeWidget = connectRecords(renderFn);
    const widget = makeWidget({});

    const helper = searchHelper({}, '', {});
    helper.search = jest.fn();

    widget.init({
      helper,
      state: helper.state,
      createURL: () => '#',
    });

    expect(renderFn).toHaveBeenLastCalledWith(
      expect.objectContaining({
        records: [],
        results: undefined,
      }),
      expect.anything()
    );

    const records = [{ fake: 'data' }, { sample: 'infos' }];

    const results = new SearchResults(helper.state, [
      { records: [].concat(records), meta: {} },
    ]);
    widget.render({
      results,
      state: helper.state,
      helper,
      createURL: () => '#',
    });

    expect(renderFn).toHaveBeenLastCalledWith(
      expect.objectContaining({
        records,
        results,
      }),
      expect.anything()
    );
  });

  it('transform items if requested', () => {
    const renderFn = jest.fn();
    const makeWidget = connectRecords(renderFn);
    const widget = makeWidget({
      transformItems: items => items.map(() => ({ name: 'transformed' })),
    });

    const helper = searchHelper({}, '', {});
    helper.search = jest.fn();

    widget.init({
      helper,
      state: helper.state,
      createURL: () => '#',
    });

    expect(renderFn).toHaveBeenNthCalledWith(
      1,
      expect.objectContaining({ records: [], results: undefined }),
      expect.anything()
    );

    const records = [{ name: 'name 1' }, { name: 'name 2' }];

    const results = new SearchResults(helper.state, [{ records, meta: {} }]);
    widget.render({
      results,
      state: helper.state,
      helper,
      createURL: () => '#',
    });

    const expectedRecords = [{ name: 'transformed' }, { name: 'transformed' }];

    expect(renderFn).toHaveBeenNthCalledWith(
      2,
      expect.objectContaining({
        records: expectedRecords,
        results,
      }),
      expect.anything()
    );
  });

  it('adds queryID if provided to results', () => {
    const renderFn = jest.fn();
    const makeWidget = connectRecords(renderFn);
    const widget = makeWidget({});

    const helper = searchHelper({}, '', {});
    helper.search = jest.fn();

    widget.init({
      helper,
      state: helper.state,
      createURL: () => '#',
    });

    const records = [{ name: 'name 1' }, { name: 'name 2' }];

    const results = new SearchResults(helper.state, [
      { records, meta: { queryID: 'theQueryID' } },
    ]);
    widget.render({
      results,
      state: helper.state,
      helper,
      createURL: () => '#',
    });

    const expectedRecords = [
      { name: 'name 1', __queryID: 'theQueryID' },
      { name: 'name 2', __queryID: 'theQueryID' },
    ];

    expect(renderFn).toHaveBeenNthCalledWith(
      2,
      expect.objectContaining({
        records: expectedRecords,
      }),
      expect.anything()
    );
  });

  describe('getWidgetSearchParameters', () => {
    it('does not add the TAG_PLACEHOLDER to the `SearchParameters` with `escapeHTML` disabled', () => {
      const render = () => {};
      const makeWidget = connectRecords(render);
      const widget = makeWidget({
        escapeHTML: false,
      });

      const actual = widget.getWidgetSearchParameters(new SearchParameters());

      expect(actual).toEqual(new SearchParameters());
    });
  });

  describe('dispose', () => {
    it('calls the unmount function', () => {
      const helper = searchHelper({}, '');

      const renderFn = () => {};
      const unmountFn = jest.fn();
      const makeWidget = connectRecords(renderFn, unmountFn);
      const widget = makeWidget();

      expect(unmountFn).toHaveBeenCalledTimes(0);

      widget.dispose({ helper, state: helper.state });

      expect(unmountFn).toHaveBeenCalledTimes(1);
    });

    it('does not throw without the unmount function', () => {
      const helper = searchHelper({}, '');

      const renderFn = () => {};
      const makeWidget = connectRecords(renderFn);
      const widget = makeWidget();

      expect(() =>
        widget.dispose({ helper, state: helper.state })
      ).not.toThrow();
    });
  });
});
