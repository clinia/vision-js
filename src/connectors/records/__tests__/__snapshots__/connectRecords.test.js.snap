// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connectRecords throws without render function 1`] = `
"The render function is not valid (received type Undefined).

See documentation: https://www.clinia.com/doc/api-reference/widgets/records/js/#connector"
`;
